python/verticox/likelihood.py

```diff
@@ -148,37 +174,52 @@ def derivative_2_diagonal(z: ArrayLike, params: Parameters, u):
 
         first_part = np.square(params.K) * np.exp(params.K * z[u]) / denominator
 
-        second_part = np.square(params.K) * np.square(np.exp(params.K * z[u])) / np.square(
-            denominator)
+        second_part = np.square(params.K) * np.square(np.exp(params.K * z[u])) / \
+                      np.square(denominator)
 
-        summed += _get_dt(t, params.Dt) * (first_part - second_part)
+        summed += params.deaths_per_t[t] * (first_part - second_part)
 
     return summed + params.K * params.rho
 
 
-def derivative_2_off_diagonal(z: ArrayLike, params: Parameters, u, v):
+@numba.njit()
+def derivative_2_off_diagonal(z: ArrayLike, params, u, v):
     min_event_time = min(params.event_times[u], params.event_times[v])
     relevant_event_times = [t for t in params.Rt.keys() if t <= min_event_time]
 
     summed = 0
 
     for t in relevant_event_times:
-        summed += _get_dt(t, params.Dt) * np.square(params.K) * np.exp(params.K * z[u]) * \
-                  np.exp(params.K * z[
-                      v]) / \
+        summed += params.deaths_per_t[t] * np.square(params.K) * np.exp(params.K * z[u]) * \
+                  np.exp(params.K * z[v]) / \
                   np.square(np.exp(params.K * z[params.Rt[t]]).sum())
 
     return -1 * summed
 
 
-def _get_dt(t, Dt):
-    # If there is no entry for time t it means that it was from a right-censored sample
-    # I'm making it an empty list so I can call len on it but it's semantically sound
-    d = Dt.get(t, [])
-    return len(d)
+@numba.njit()
+def hessian_parametrized(z: types.float64[:], params: Parameters):
+    # The hessian is a N x N matrix where N is the number of elements in z
+    N = z.shape[0]
+    mat = np.zeros((N, N))
 
+    for u in range(N):
```

Comment:
Will your servers have multiple cores? This NxN matrix calc can have every value computed independently of the others, so if you wanted to parallelise anything I would do it here. This is also an inner loop call, if I'm not mistaken

Reply:
You're right. I changed the jit thing to `@njit(parallel=True)`

Reply:
I suppose you then need to use `prange` instead of `range`?

Reply:
parallelizing only slow things down unfortunately :(
I don't know why this is. I tried to parallelize the creating of the hessian and jacobian matrix but it doesn't help at all

Reply:
Perhaps trying `prange` only on the outer loop could help (by increasing load on each core, reducing overhead of parallelisation)